From 02b544592e61a27a32acae2ec028ef583b57fbdc Mon Sep 17 00:00:00 2001
From: Petros Angelatos <petrosagg@gmail.com>
Date: Thu, 3 Nov 2016 00:38:14 +0000
Subject: [PATCH] graph: aufs: durably write layer on disk before returning

This patch makes sure the layer contents are synced to disk before
reporting the ApplyDiff operation as successful. This prevents
/var/lib/docker corruption but the method used here is not the most
efficient since it will sync all the currently mounted filesystems.

Signed-off-by: Petros Angelatos <petrosagg@gmail.com>
---
 daemon/graphdriver/aufs/aufs.go | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/daemon/graphdriver/aufs/aufs.go b/daemon/graphdriver/aufs/aufs.go
index 51054fa..50f8197 100644
--- a/daemon/graphdriver/aufs/aufs.go
+++ b/daemon/graphdriver/aufs/aufs.go
@@ -397,6 +397,12 @@ func (a *Driver) ApplyDiff(id, parent string, diff archive.Reader) (size int64,
 		return
 	}
 
+	// FIXME: Instead of syncing all the filesystems we should be fsyncing each
+	// file as the tar archive gets unpacked
+	if err = syscall.Sync(); err != nil {
+		return
+	}
+
 	return a.DiffSize(id, parent)
 }
 
-- 
2.10.2

